### PR TITLE
Word comparator ( MD-Jubayer-Khan_2nd )

### PR DESCRIPTION
**Pull request by MD-Jubayer-Khan**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_word_comparator.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class TestWordComparator(unittest.TestCase):
+    """
+    Unit tests for the are_words_same function.
+    """
+
+    def test_same_words(self):
+        """It should return True for words with the same sorted characters."""
+        self.assertTrue(are_words_same("listen", "silent"))
+        self.assertTrue(are_words_same("word", "wrdo"))
```
Comment (MuqadsaT) [inline solutions/tests/test_word_comparator.py]: The test_same_words and test_different_words tests have two assertions each. It's generally better to have separate tests for each scenario.

File: solutions/tests/test_word_comparator.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def test_case_sensitivity(self):
+        """It should consider case sensitivity while comparing."""
+        self.assertFalse(are_words_same("Listen", "silent"))
+        self.assertFalse(are_words_same("Word", "word"))
+
+    def test_partial_overlap(self):
+        """It should return False for words with some common characters."""
+        self.assertFalse(are_words_same("abcd", "abc"))
+
```
Comment (MuqadsaT) [inline solutions/tests/test_word_comparator.py:91]: Add tests for defensive assertions: Include tests to check the function's behavior with non-string inputs (e.g., integers, lists, None).

File: solutions/tests/test_word_comparator.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+
+from word_comparator import are_words_same
+
+
+class TestWordComparator(unittest.TestCase):
+    """
+    Unit tests for the are_words_same function.
+    """
```
Comment (MuqadsaT) [inline solutions/tests/test_word_comparator.py:20]: Add tests for boundary cases: Include tests for very long strings, strings with special characters (e.g., punctuation, whitespace), and potentially very short strings (e.g., one-character strings).

File: solutions/word_comparator.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    """
+    Compares two words by sorting their characters and checking for equality.
+
+    Parameters:
+        word1 (str): The first word to compare.
+        word2 (str): The second word to compare.
+
+    Returns:
+        bool: True if the sorted characters of both words are the same, False otherwise.
+
```
Comment (MuqadsaT) [inline solutions/word_comparator.py:22]: This function doesn't have any explicit assertions.

File: solutions/word_comparator.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        >>> are_words_same("word", "wrdo")
+        True
+
+        >>> are_words_same("hello", "world")
+        False
+
+        >>> are_words_same("", "")
+        True
+    """
+    return sorted(word1) == sorted(word2)
```
Comment (MuqadsaT) [inline solutions/word_comparator.py:30]: assert isinstance(word1, str), "word1 must be a string"
assert isinstance(word2, str), "word2 must be a string"

File: solutions/word_comparator.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+Description: This module provides a function to compare two words
+by checking if their sorted characters are the same.
+
+Created on: 03 01 25
+@author: MD Jubayer Khan
+"""
+
+
+def are_words_same(word1: str, word2: str) -> bool:
```
Comment (MuqadsaT) [inline solutions/word_comparator.py]: The function name (are_words_same) and file name (word_comparator.py) are descriptive and convey the function's purpose. For consistency within the project, consider aligning the file name more closely with the function name. For example, are_words_same.py would be more consistent with the function name.

File: solutions/word_comparator.py (hunk 5 above)
Comment (MuqadsaT) [inline solutions/word_comparator.py:30]: A comment like below can further clarify the strategy:
Sort the characters of each word and compare the resulting lists.

**Review by MuqadsaT — CHANGES_REQUESTED** [review]

**Comment by MD-Jubayer-Khan** [comment]
ok. I'll change them

**Review by MD-Jubayer-Khan — COMMENTED** [review]
Modified. Let me know if there needs any further modification

File: solutions/word_comparator.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        word1 (str): The first word to compare.
+        word2 (str): The second word to compare.
+
+    Returns:
+        bool: True if the sorted characters of both words are the same, False otherwise.
+
+    Raises:
+        TypeError: If either input is not a string.
+    """
+    if not isinstance(word1, str) or not isinstance(word2, str):
```
Comment (MuqadsaT) [inline solutions/word_comparator.py]: The assertion if not isinstance(word1, str) or not isinstance(word2, str): includes logic within the assertion itself.
Ideally, assertions should be concise and clearly express a single assumption.
Consider splitting this into two separate assertions

**Review by MuqadsaT — DISMISSED** [review]

**Review by MuqadsaT — APPROVED** [review]

**Comment by MuqadsaT** [comment]
This code includes comprehensive docstrings, unit tests, and adheres to best practices for readability and maintainability. Good work @MD-Jubayer-Khan  🚀
